src/make_particle.c

```diff
@@ -2211,6 +2271,38 @@ void MakeParticle(void)
 				 */
 				if ((yj==0 || yj==-jagged) && (zj==0 || zj==-jagged)) mat=0;
 				break;
+			case SH_ONION:
+				r2=xr*xr+yr*yr+zr*zr;
+				// only consider dipoles inside the sphere				
+				if (r2<=0.25) {
+					// first test if inside the core, then test if in layers from the outside in
+					if (r2<=onion_r2[nlayers-2]) mat=nlayers-1; // inside core 
+					else {
+						for (layer_ctr=0;layer_ctr<(nlayers-1);layer_ctr++) {
+							if (r2>onion_r2[layer_ctr]) {
+								mat=layer_ctr;
+								break;
+							}
```

Comment:
Have you considered any acceleration for such search? 

Actually, I tried running it with 250 layers, and the time for make_particle was always at least 10 times smaller that a single iteration. So definitely not a bottleneck. However, the code may even be neater if bsearch function is used.

Reply:
Respectfully, I'm not sure that using `bsearch` would significantly improve performance (given that `make_particle` is not an overall bottleneck in running ADDA) or would make this easier to read. Unless I'm missing something, `bsearch` would find an exact match between the `r2` value of a given dipole and an element of `onion_r2`, but that's not what is needed here. Moreover, using `bsearch` would require writing a comparison function.